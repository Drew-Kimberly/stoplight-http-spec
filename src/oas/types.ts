import {
  IApiKeySecurityScheme,
  IBasicSecurityScheme,
  IBearerSecurityScheme,
  IOauth2SecurityScheme,
  IOpenIdConnectSecurityScheme,
} from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export declare type HttpSecurityScheme =
  | IApiKeySecurityScheme
  | IBearerSecurityScheme
  | IBasicSecurityScheme
  | IOauth2SecurityScheme
  | IOpenIdConnectSecurityScheme;

// prefixing with T to avoid tslinter "no-shadow" error
export interface ITranslateOperationOptions<TSpec> {
  document: TSpec;
  path: string;
  method: string;
}

export type Oas2TranslateOperationOptions = ITranslateOperationOptions<Spec>;
export type Oas3TranslateOperationOptions = ITranslateOperationOptions<OpenAPIObject>;

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}
