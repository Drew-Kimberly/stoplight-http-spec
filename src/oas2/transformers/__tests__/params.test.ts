import { HttpParamStyles } from '@stoplight/types';
import { QueryParameter } from 'swagger-schema-official';

import {
  translateFromFormDataParameter,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToHeaderParams,
  translateToPathParameter,
  translateToQueryParameter,
} from '../params';

describe('params.translator', () => {
  const consumes = ['*'];

  describe('translateToHeaderParam', () => {
    test('should translate header param', () => {
      expect(
        translateToHeaderParam({
          required: true,
          name: 'name',
          type: 'integer',
          in: 'header',
          minimum: 12,
          description: 'desc',
        }),
      ).toMatchSnapshot();
    });
  });

  describe('translateToHeaderParams', () => {
    test('should translate empty dictionary to empty array', () => {
      expect(translateToHeaderParams({})).toMatchSnapshot();
    });

    test('should translate to simple header param', () => {
      expect(
        translateToHeaderParams({
          'header-name': {
            description: 'a description',
            type: 'string',
          },
        }),
      ).toMatchSnapshot();
    });

    test('should translate to multiple header params', () => {
      expect(
        translateToHeaderParams({
          'header-name': {
            description: 'a description',
            type: 'string',
          },
          'plain-tex': {
            description: 'another description',
            type: 'string',
          },
        }),
      ).toMatchSnapshot();
    });
  });

  describe('translateToBodyParameter', () => {
    test('should translate to body parameter', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              format: 'e-mail',
            },
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });

    test('given x-examples should translate to body parameter with multiple examples', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              format: 'e-mail',
            },
            // @ts-ignore: "x-examples" isn't technically supported by the OAS2 spec
            'x-examples': {
              'example-1': {
                hello: 'world',
              },
              'example-2': {
                foo: 'bar',
              },
            },
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('translateToFormDataParameter', () => {
    test('given request body with empty encodings should create encoding', () => {
      expect(
        translateFromFormDataParameter(
          {
            name: 'name',
            type: 'string',
            description: 'desc',
            required: true,
            in: 'formData',
          },
          {
            contents: [
              {
                mediaType: 'application/json',
              },
            ],
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });

    test('given request body with existing encoding should append', () => {
      expect(
        translateFromFormDataParameter(
          {
            name: 'name',
            type: 'string',
            description: 'desc',
            required: true,
            in: 'formData',
          },
          {
            contents: [
              {
                examples: [],
                mediaType: 'application/json',
                encodings: [
                  {
                    headers: [],
                    property: 'prop',
                    style: HttpParamStyles.SpaceDelimited,
                  },
                ],
              },
            ],
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });

    test('given no request body should create one', () => {
      expect(
        translateFromFormDataParameter(
          {
            name: 'name',
            type: 'string',
            description: 'desc',
            required: true,
            in: 'formData',
          },
          null,
          consumes,
        ),
      ).toMatchSnapshot();
    });

    test('given a request body it should add a property to the schema', () => {
      expect(
        translateFromFormDataParameter(
          {
            name: 'name',
            type: 'string',
            description: 'desc',
            required: true,
            in: 'formData',
          },
          {
            contents: [
              {
                examples: [],
                mediaType: 'application/json',
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'number' },
                    count: { type: 'number' },
                  },
                  required: ['id'],
                },
              },
            ],
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('translateToQueryParameter', () => {
    const parameter: QueryParameter = {
      required: true,
      description: 'descr',
      name: 'name',
      in: 'query',
      type: 'number',
      allowEmptyValue: true,
    };

    test.each(['pipes', 'ssv', 'csv', 'multi', 'something-else'])('translate style: %s', style => {
      expect(translateToQueryParameter({ ...parameter, collectionFormat: style } as QueryParameter)).toMatchSnapshot();
    });
  });

  describe('translateToPathParameter', () => {
    test('should translate', () => {
      expect(
        translateToPathParameter({
          required: true,
          description: 'descr',
          name: 'name',
          in: 'path',
          type: 'string',
        }),
      ).toMatchSnapshot();
    });
  });
});
