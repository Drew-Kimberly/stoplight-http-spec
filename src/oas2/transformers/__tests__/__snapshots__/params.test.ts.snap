// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`params.translator translateToBodyParameter given x-examples should translate to body parameter with multiple examples 1`] = `
Object {
  "contents": Array [
    Object {
      "examples": Array [
        Object {
          "key": "example-1",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "key": "example-2",
          "value": Object {
            "foo": "bar",
          },
        },
      ],
      "mediaType": "*",
      "schema": Object {
        "format": "e-mail",
      },
    },
  ],
  "description": "descr",
  "required": true,
}
`;

exports[`params.translator translateToBodyParameter should translate to body parameter 1`] = `
Object {
  "contents": Array [
    Object {
      "examples": Array [],
      "mediaType": "*",
      "schema": Object {
        "format": "e-mail",
      },
    },
  ],
  "description": "descr",
  "required": true,
}
`;

exports[`params.translator translateToFormDataParameter given a request body it should add a property to the schema 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "examples": Array [],
      "mediaType": "application/json",
      "schema": Object {
        "properties": Object {
          "count": Object {
            "type": "number",
          },
          "id": Object {
            "type": "number",
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
    },
  ],
}
`;

exports[`params.translator translateToFormDataParameter given no request body should create one 1`] = `
Object {
  "allowEmptyValue": undefined,
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "mediaType": "*",
      "schema": Object {
        "description": "desc",
        "type": "string",
      },
    },
  ],
}
`;

exports[`params.translator translateToFormDataParameter given request body with empty encodings should create encoding 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "mediaType": "application/json",
      "schema": Object {
        "description": "desc",
        "type": "string",
      },
    },
  ],
}
`;

exports[`params.translator translateToFormDataParameter given request body with existing encoding should append 1`] = `
Object {
  "contents": Array [
    Object {
      "encodings": Array [
        Object {
          "headers": Array [],
          "property": "prop",
          "style": "spaceDelimited",
        },
        Object {
          "property": "name",
          "style": "form",
        },
      ],
      "examples": Array [],
      "mediaType": "application/json",
      "schema": Object {
        "description": "desc",
        "type": "string",
      },
    },
  ],
}
`;

exports[`params.translator translateToHeaderParam should translate header param 1`] = `
Object {
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "desc",
    "minimum": 12,
    "type": "integer",
  },
  "style": "simple",
}
`;

exports[`params.translator translateToHeaderParams should translate empty dictionary to empty array 1`] = `Array []`;

exports[`params.translator translateToHeaderParams should translate to multiple header params 1`] = `
Array [
  Object {
    "name": "header-name",
    "schema": Object {
      "description": "a description",
      "type": "string",
    },
    "style": "simple",
  },
  Object {
    "name": "plain-tex",
    "schema": Object {
      "description": "another description",
      "type": "string",
    },
    "style": "simple",
  },
]
`;

exports[`params.translator translateToHeaderParams should translate to simple header param 1`] = `
Array [
  Object {
    "name": "header-name",
    "schema": Object {
      "description": "a description",
      "type": "string",
    },
    "style": "simple",
  },
]
`;

exports[`params.translator translateToPathParameter should translate 1`] = `
Object {
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "type": "string",
  },
  "style": "simple",
}
`;

exports[`params.translator translateToQueryParameter translate style: csv 1`] = `
Object {
  "allowEmptyValue": true,
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "minLength": 1,
    "type": "number",
  },
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: multi 1`] = `
Object {
  "allowEmptyValue": true,
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "minLength": 1,
    "type": "number",
  },
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: pipes 1`] = `
Object {
  "allowEmptyValue": true,
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "minLength": 1,
    "type": "number",
  },
  "style": "pipeDelimited",
}
`;

exports[`params.translator translateToQueryParameter translate style: something-else 1`] = `
Object {
  "allowEmptyValue": true,
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "minLength": 1,
    "type": "number",
  },
  "style": "form",
}
`;

exports[`params.translator translateToQueryParameter translate style: ssv 1`] = `
Object {
  "allowEmptyValue": true,
  "name": "name",
  "required": true,
  "schema": Object {
    "description": "descr",
    "minLength": 1,
    "type": "number",
  },
  "style": "spaceDelimited",
}
`;
